pkg/mmesh/etcd_secret.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2021 IBM Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package mmesh
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	EtcdConnectionKey = "etcd_connection"
+)
+
+// An EtcdSecret represents the etcd configuation for a user namespace
+type EtcdSecret struct {
+	Log                 logr.Logger
+	Name                string
+	Namespace           string
+	ControllerNamespace string
+	EtcdConfig          *EtcdConfig
+	Scheme              *runtime.Scheme
+}
+
+func (es EtcdSecret) Apply(ctx context.Context, cl client.Client) error {
+	commonLabelValue := "modelmesh-controller"
+
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      es.Name,
+			Namespace: es.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": commonLabelValue,
+			},
+		},
+	}
+	err := es.addData(s)
+	if err != nil {
+		return err
+	}
+	err = cl.Create(ctx, s)
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = cl.Update(ctx, s)
+	}
```

Comment:
The typical state will be already created, so this will cause some unnecessary churn... would be best to do a Get first above before adding the data (which will come from local cache), and then create only if it doesn't exist. If it _does_ exist then just update (or even better check first to see if the data already matches in which case there's nothing to do)